tiledb/sm/query/query_remote_buffer_storage.cc

```diff
@@ -0,0 +1,277 @@
+    uint64_t cache_buffer_cells =
+        cache.second.buffer_.size() / cache.second.cell_size_;
+
+    if (buffer_cells + cache_buffer_cells <= cell_num_per_tile_) {
+      cache_write = true;
+    }
+  }
+
+  return cache_write;
+}
+
+void QueryRemoteBufferStorage::cache_write() {
+  for (const auto& name : query_->buffer_names()) {
+    auto query_buffer = query_->buffer(name);
+    bool is_var = query_->array_schema().var_size(name);
+    bool is_nullable = query_->array_schema().is_nullable(name);
+    auto& cache = caches_[name];
+
+    // Buffer element size, or size of offsets if attribute is var-sized.
+    uint64_t buffer_cells = *query_buffer.buffer_size_ / cache.cell_size_;
+    uint64_t buffer_cache_bytes = buffer_cells * cache.cell_size_;
+
+    // Write fixed attribute data into cache buffer.
+    throw_if_not_ok(
+        cache.buffer_.write(query_buffer.buffer_, buffer_cache_bytes));
+    *query_buffer.buffer_size_ -= buffer_cache_bytes;
+
+    if (is_var) {
+      // Adjust appended offsets to be relative to already cached data.
+      cache.adjust_offsets(buffer_cache_bytes);
+
+      // Write the variable sized data into cache.
+      throw_if_not_ok(cache.buffer_var_.write(
+          query_buffer.buffer_var_, *query_buffer.buffer_var_size_));
+      *query_buffer.buffer_var_size_ -= *query_buffer.buffer_var_size_;
+    }
+
+    if (is_nullable) {
+      // Write the validity buffers into cache.
+      throw_if_not_ok(cache.buffer_validity_.write(
+          query_buffer.validity_vector_.buffer(), buffer_cells));
+      *query_buffer.validity_vector_.buffer_size() -= buffer_cells;
+    }
+  }
+}
+
+void QueryRemoteBufferStorage::make_buffers_tile_aligned() {
+  // Only applicable for remote global order writes
+  if (query_->type() != QueryType::WRITE ||
+      query_->layout() != Layout::GLOBAL_ORDER) {
+    return;
+  }
+
+  for (auto& cache : caches_) {
+    const auto& query_buffer = query_->buffer(cache.first);
+    bool is_var = query_->array_schema().var_size(cache.first);
+    bool is_nullable = query_->array_schema().is_nullable(cache.first);
+
+    // Buffer element size, or size of offsets if attribute is var-sized.
+    uint64_t buffer_cells =
+        *query_buffer.buffer_size_ / cache.second.cell_size_;
+    uint64_t cache_buffer_cells =
+        cache.second.buffer_.size() / cache.second.cell_size_;
+    uint64_t total_buffer_cells = buffer_cells + cache_buffer_cells;
+
+    // If buffer_cells does not fill a tile, cache the entire buffer.
+    uint64_t tile_cache_cells = buffer_cells;
+    uint64_t buffer_cache_bytes = tile_cache_cells * cache.second.cell_size_;
+    // Check if data between cache and user buffer is tile-aligned.
+    if (total_buffer_cells >= cell_num_per_tile_ &&
+        total_buffer_cells % cell_num_per_tile_ == 0) {
+      // Submit all data and cache nothing after submit.
+      cache.second.cache_bytes_ = 0;
+      continue;
+    }
+
+    if (buffer_cells >= cell_num_per_tile_) {
```

Comment:
I'm not sure I get this... Shouldn't this always be: 
buffer_cache_bytes = (total_buffer_cells % cell_num_per_tile_) * cache.second.cell_size_;